src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -136,66 +110,92 @@ private enum BatchWriterState
             Error
         }
 
+        /// <summary>The message for the operation that is currently written; or null if no operation is written right now.</summary>
+        internal ODataBatchOperationMessage CurrentOperationMessage
+        {
+            get
+            {
+                Debug.Assert(this.currentOperationRequestMessage == null || this.currentOperationResponseMessage == null, "Only request or reponse message can be set, not both.");
+                if (this.currentOperationRequestMessage != null)
+                {
+                    Debug.Assert(!this.outputContext.WritingResponse, "Request message can only be set when writing request.");
+                    return this.currentOperationRequestMessage.OperationMessage;
+                }
+                else if (this.currentOperationResponseMessage != null)
+                {
+                    Debug.Assert(this.outputContext.WritingResponse, "Response message can only be set when writing response.");
+                    return this.currentOperationResponseMessage.OperationMessage;
+                }
+                else
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets or Sets the content Id of the current operation.
+        /// </summary>
+        protected string CurrentOperationContentId
+        {
+            get { return this.currentOperationContentId; }
+            set { this.currentOperationContentId = value; }
+        }
+
+        /// <summary>
+        /// Gets or Sets the batch writer's state.
+        /// </summary>
+        protected BatchWriterState State
+        {
+            get { return this.state; }
+            set { this.state = value; }

```

Comment:
> set { this.state = value; } [](start = 11, length = 28)

I would generally expect the state to be set within the ODataBatchWriter base class as the derived classed called methods for each transition.